File: py/achkit/fields.py
"""Fixed-width NACHA field primitives. Every record is exactly 94 characters."""
from __future__ import annotations

RECORD_LEN = 94
BLOCKING_FACTOR = 10


class AchError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


def _ascii(s: str) -> str:
    return "".join(c if 0x20 <= ord(c) <= 0x7E else " " for c in s)


def alnum(v, length: int) -> str:
    """Left-justified, space-padded, upper-cased alphanumeric field."""
    s = _ascii(str("" if v is None else v).upper())
    return s[:length] if len(s) >= length else s + " " * (length - len(s))


def num(v, length: int) -> str:
    """Right-justified, zero-padded numeric field. Raises on overflow."""
    s = "".join(c for c in str(0 if v is None else v) if c.isdigit())
    if len(s) > length:
        raise AchError("FIELD_OVERFLOW", f"numeric field overflow: {s} > {length} digits")
    return s.rjust(length, "0")


def money(cents: int, length: int) -> str:
    if not isinstance(cents, int) or cents < 0:
        raise AchError("BAD_AMOUNT", f"amount must be a non-negative integer of cents, got {cents}")
    return num(cents, length)


def digits(v: str) -> str:
    return "".join(c for c in str(v) if c.isdigit())

File: js/src/fields.ts
// Fixed-width NACHA field primitives. Every record is exactly 94 characters;
// these helpers guarantee width and justification so callers never count bytes.

export const RECORD_LEN = 94
export const BLOCKING_FACTOR = 10

const ASCII = /[^\x20-\x7E]/g

/** Left-justified, space-padded, upper-cased alphanumeric field. Truncates to len. */
export function alnum(v: string | number | undefined, len: number): string {
  const s = String(v ?? '')
    .toUpperCase()
    .replace(ASCII, ' ')
  return s.length >= len ? s.slice(0, len) : s + ' '.repeat(len - s.length)
}

/** Right-justified, zero-padded numeric field. Throws if the value overflows len. */
export function num(v: string | number | undefined, len: number): string {
  const s = String(v ?? 0).replace(/[^0-9]/g, '')
  if (s.length > len) throw new AchError('FIELD_OVERFLOW', `numeric field overflow: ${s} > ${len} digits`)
  return s.padStart(len, '0')
}

/** Money in integer cents rendered as a right-justified zero-padded field. */
export function money(cents: number, len: number): string {
  if (!Number.isInteger(cents) || cents < 0) throw new AchError('BAD_AMOUNT', `amount must be a non-negative integer of cents, got ${cents}`)
  return num(cents, len)
}

export function digits(v: string): string {
  return String(v).replace(/[^0-9]/g, '')
}

export class AchError extends Error {
  code: string
  constructor(code: string, message: string) {
    super(message)
    this.name = 'AchError'
    this.code = code
  }
}
